### ISSUE TITLE
0xpep7 - Hard Fork Replay Vulnerability in Contract Initialization due to Hardcoded chainId

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2023-12-dodo-gsp-judging#54]
0xpep7

medium

# Hard Fork Replay Vulnerability in Contract Initialization due to Hardcoded chainId

## Summary
The contract initialization process poses a vulnerability related to hard forks, potentially leading to replay attacks due to the incorrect handling of the chainId parameter in the DOMAIN_SEPARATOR.

## Vulnerability Detail
The root cause of the vulnerability is that the chainId parameter, crucial for constructing the DOMAIN_SEPARATOR, is computed during initialization (init) and then hardcoded. This means that even after a hard fork, the DOMAIN_SEPARATOR value remains the same, pointing to an incorrect chainId. Furthermore, the init function can only be called once, making it impossible to update the value after a hard fork.

## Impact
In the event of a hard fork on the operating network, the DOMAIN_SEPARATOR value becomes invalid, and this could open up the possibility of replay attacks. While the impact is contingent on external factors such as a hard fork, it is a medium-severity issue that could have significant consequences in certain scenarios.

## Code Snippet
https://github.com/sherlock-audit/2023-12-dodo-gsp/blob/main/dodo-gassaving-pool/contracts/GasSavingPool/impl/GSP.sol#L92

## Tool used
Manual Review

## Recommendation
To mitigate this vulnerability, the DOMAIN_SEPARATOR variable should be recomputed every time by placing the current value of chainId dynamically during runtime. This ensures that the contract remains adaptable to changes in the chainId due to hard forks.

Duplicate of #113
